src/imitation/util/logger.py

```diff
@@ -0,0 +1,197 @@
+
+  def __init__(self,
+               default_logger: sb_logger.Logger,
+               format_strs: Sequence[str] = ('stdout', 'log', 'csv')):
+    """A logger with a context for accumulating mean values.
+
+    Args:
+      default_logger: The default logger when not in the a `accumulate_means`
+        context. Also the logger to which mean values are written to when
+        contexts are over.
+      format_strs: An list of output format strings that should be used by
+        every Logger initialized by this class during an `AccumulatingMeans`
+        context. For details on available output formats see
+        `stable_baselines.logger.make_output_format`.
+    """
+    self.default_logger = default_logger
+    self.current_logger = None
+    self._cached_loggers = {}
+    self._subdir = None
+    self.format_strs = format_strs
+    super().__init__(folder=self.default_logger.dir, output_formats=None)
+
+  @contextlib.contextmanager
+  def accumulate_means(self, subdir: str):
+    """Temporarily modifies this _HierarchicalLogger to accumulate means values.
+
+    During this context, `self.logkv(key, value)` writes the "raw" values in
+    "{self.default_logger.log_dir}/{subdir}" under the key "raw/{subdir}/{key}".
+    At the same time, any call to `self.logkv` will also accumulate mean values
+    on the default logger by calling
+    `self.default_logger.logkv_mean(f"mean/{subdir}/{key}", value)`.
+
+    During the context, `self.logkv(key, value)` will write the "raw" values in
+    `"{self.default_logger.log_dir}/subdir"` under the key "raw/{subdir}/key".
+
+    After the context exits, calling `self.dumpkvs()` will write the means
+    of all the "raw" values accumulated during this context to
+    `self.default_logger` under keys with the prefix `mean/{subdir}/`
+
+    Args:
+      subdir: A string key which determines the `folder` where raw data is
+        written and temporary logging prefixes for raw and mean data. Entering
+        an `accumulate_means` context in the future with the same `subdir`
+        will safely append to logs written in this folder rather than
+        overwrite.
+    """
+    if self.current_logger is not None:
+      raise RuntimeError("Nested `accumulate_means` context")
+
+    if subdir in self._cached_loggers:
+      logger = self._cached_loggers[subdir]
+    else:
+      folder = os.path.join(self.default_logger.dir, "raw", subdir)
+      os.makedirs(folder)
+      output_formats = _build_output_formats(folder, self.format_strs)
+      logger = sb_logger.Logger(folder, output_formats)
+      self._cached_loggers[subdir] = logger
+
+    try:
+      self.current_logger = logger
+      self._subdir = subdir
+      yield
+    finally:
+      self.current_logger = None
+      self._subdir = None
+
+  def logkv(self, key, val):
+    if self.current_logger is not None:
+      assert self._subdir is not None
+      raw_key = os.path.join("raw", self._subdir, key)
```

Comment:
Maybe this is bike-shedding a little, but is it really necessary to have the `raw/{subdir}/` prefix on all of the entries in the `self.current_logger` log? As I understand it we'll end up with a bunch of log files in which _all log entries_ start with `raw/subdir/`, which would be redundant given that you can infer what kind of entries they are from the directory hierarchy (i.e. the raw log file is already in a `raw/subdir` folder).

(overall I still like the design!)

Reply:
Adam thought this was weird too, we had a comment thread about this [here](https://github.com/HumanCompatibleAI/imitation/pull/135#discussion_r362351049).

Reasons for redundant-looking prefix:
(1) So that in Tensorboard plots, mean plots are automatically separated from raw plots (which have a different x axis). (TB plots are grouped by key name). 
(2) To distinguish between raw and mean keys in `stdout` logging (otherwise they look identical).

(1) is my primary motivation. Adam was convinced by (2). Though now that I think about it, it might be better to just turn off raw stdout logs and show mean stdout logs only.